vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -795,10 +821,20 @@ def done_callback(f: Future[dict[int, str]], eid=remote_engine_id):
 
             fut.add_done_callback(done_callback)
 
-        # TODO: handle failure state of future in the
-        # callback, we want to fail the request in this case.
+        # check handshake success before proceeding with request
         def request_ready(_f: Future[Any], entry=(req_id, meta)):
-            self._ready_requests.put(entry)
+            try:
+                # check if handshake succeeded
+                _f.result()
+                self._ready_requests.put(entry)
+            except Exception:
+                # handshake failed - mark blocks as invalid
+                logger.error(
+                    "Handshake failed for request %s, marking blocks as invalid", req_id
+                )
+                req_meta = self._recving_metadata.get(req_id)
+                if req_meta:
+                    self._invalid_block_ids.update(req_meta.local_block_ids)
```

Comment:
nit: compactness suggestion
```suggestion
                if req_meta := self._recving_metadata.get(req_id):
                    self._invalid_block_ids.update(req_meta.local_block_ids)
```